contracts/contract_call.go

```diff
@@ -0,0 +1,120 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/common/hexutil"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/log"
+)
+
+// Method represents a contract's method
+type Method struct {
+	abi    *abi.ABI
+	method string
+	maxGas uint64
+}
+
+// NewMethod creates a new contract message
+func NewMethod(abi *abi.ABI, method string, maxGas uint64) Method {
+	return Method{
+		abi:    abi,
+		method: method,
+		maxGas: maxGas,
+	}
+}
+
+func (am Method) Bind(contractAddress common.Address) *BoundMethod {
+	return &BoundMethod{
+		Method:         am,
+		resolveAddress: noopResolver(contractAddress),
+	}
+}
+
+// encodeCall will encodes the msg into []byte format for EVM consumption
+func (am Method) encodeCall(args ...interface{}) ([]byte, error) {
+	return am.abi.Pack(am.method, args...)
+}
+
+// decodeResult will decode the result of msg execution into the result parameter
+func (am Method) decodeResult(result interface{}, output []byte) error {
+	if result == nil {
+		return nil
+	}
+	return am.abi.Unpack(result, am.method, output)
+}
+
+func NewBoundMethod(contractAddress common.Address, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return NewMethod(abi, methodName, maxGas).Bind(contractAddress)
+}
+
+func NewRegistryContractMethod(registryId common.Hash, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return &BoundMethod{
+		Method: NewMethod(abi, methodName, maxGas),
+		resolveAddress: func(caller vm.EVMCaller) (common.Address, error) {
+			return resolveAddressForCall(caller, registryId, methodName)
+		},
+	}
+}
+
+type BoundMethod struct {
+	Method
+	resolveAddress func(vm.EVMCaller) (common.Address, error)
+}
+
+func (bm *BoundMethod) VMQuery(caller vm.EVMCaller, result interface{}, args ...interface{}) error {
+	return bm.Query(caller, result, VMAddress, args...)
+}
+
+func (bm *BoundMethod) Query(caller vm.EVMCaller, result interface{}, sender common.Address, args ...interface{}) error {
+	return bm.run(caller, result, true, sender, nil, args...)
+}
+
+func (bm *BoundMethod) VMExecute(caller vm.EVMCaller, result interface{}, value *big.Int, args ...interface{}) error {
+	return bm.run(caller, result, false, VMAddress, value, args...)
+}
+
+func (bm *BoundMethod) Execute(caller vm.EVMCaller, result interface{}, sender common.Address, value *big.Int, args ...interface{}) error {
+	return bm.run(caller, result, false, sender, value, args...)
+}
```

Comment:
I think there might be a problem here, because you are passing `sender` but the from address has been hardcoded in the `EVMCaller` constructor as the zero address. I'd be in favoiur of getting rid of both `Query` and `Execute` as they do not seem to be used in either this or the subsequent PR.

```go
func NewEVMCaller(chain ExtendedChainContext, header *types.Header, state vm.StateDB) vm.EVMCaller {
	// The EVM Context requires a msg, but the actual field values don't really matter for this case.
	// Putting in zero values.
	context := New(common.ZeroAddress, common.Big0, header, chain, nil)

	return &evmCaller{
		vmConfig:    *chain.GetVMConfig(),
		vmContext:   context,
		chainConfig: chain.Config(),
		state:       state,
	}
}
```

Reply:
yeah, good point. I rediscoverd that yesterday. So evm "origin" is hardcoded to zeroAddress and then we do set the sender (`msg.from` in solitidy). That's nasty, we should use the same.

Only the tobin tax call changes the sender... and i'd argue it should. So i'll change it to always use the `vmaddress`

In fact, how do you feel about renaming `EVMCaller` to `SystemCaller` :D?

Reply:
:D I think actually `SystemEVM` would be a better name here `SystemCaller` has no reference to EVM and given the generic use of the term `call` in programming and also the connotation of `System` with OS I think it would be easy to misinterpret.

Reply:
Happy to do the change, but just want to make sure we agree on the name first

Reply:
chaging it to SystemEVM then

Reply:
For now, i've renamed to `SystemEVM` but kept the `VMQUery` etc... instead i'm creating the vm.Context on the fly using the sender, instead of zeroAddress